File: recover-the-original-array.cc
#define ALL(x) (x).begin(), (x).end()
#define FOR(i, a, b) for (long i = (a); i < (b); i++)
#define REP(i, n) for (long i = 0; i < (n); i++)

class Solution {
public:
  vector<int> recoverArray(vector<int>& a) {
    sort(ALL(a));
    int n = a.size();
    vector<char> used(n);
    auto recover = [&](int d) {
      vector<int> res;
      int j = 0;
      REP(i, n) {
        if (used[i]) continue;
        while (j < n && a[j] < a[i]+d) j++;
        if (!(j < n && a[j] == a[i]+d)) break;
        used[j++] = 1;
        res.push_back(a[i]+d/2);
      }
      if (res.size() != n/2) res.clear();
      while (j) used[--j] = 0;
      return res;
    };

    FOR(i, 1, n) {
      int d = a[i]-a[0];
      if (d == 0 || d % 2) continue;
      vector<int> res = recover(d);
      if (res.size())
        return res;
    }
    return {};
  }
};

File: recover-the-original-array.nim
# Recover the Original Array
import std/[algorithm, sequtils]

type cintA = ptr UncheckedArray[cint]
proc malloc(size: csize_t): pointer {.importc.}

proc recover(a: seq[cint], used: var seq[int8], d: cint): seq[cint] {.exportc.} =
  var res: seq[cint]
  var j = 0
  let n = a.len
  for i in 0..<n:
    let x = a[i]
    if used[i] != 0: continue
    while j < n and a[j] < a[i]+d:
      j += 1
    if not (j < n and a[j] == a[i]+d): break
    used[j] = 1
    j += 1
    res.add x+(d shr 1)
  if res.len * 2 != n:
    res.setLen 0
  while j > 0:
    j -= 1
    used[j] = 0
  return res

proc recoverArray(nums: cintA, n: int, returnSize: ptr cint): cintA {.exportc.} =
  var a = nums.toOpenArray(0, n-1).toSeq
  var used = newSeq[int8](n)
  a.sort
  for i in 1..<n:
    let d = a[i]-a[0]
    if d == 0 or d mod 2 == 1: continue
    var res = recover(a, used, d)
    if res.len > 0:
      returnSize[] = res.len.cint
      result = cast[cintA](malloc((cint.sizeof * res.len).csize_t))
      for i in 0..<res.len:
        result[i] = res[i]
      break
